src/log_unstable.rs

```diff
@@ -85,29 +85,19 @@ impl Unstable {
         }
     }
 
-    /// Moves the stable offset up to the index. Provided that the index
-    /// is in the same election term.
-    pub fn stable_to(&mut self, idx: u64, term: u64) {
-        let t = self.maybe_term(idx);
-        if t.is_none() {
-            return;
-        }
-
-        if t.unwrap() == term && idx >= self.offset {
-            let start = idx + 1 - self.offset;
-            self.entries.drain(..start as usize);
-            self.offset = idx + 1;
+    /// Returns the unstable entries and moves the stable offset up to the
+    /// last index, if there is any.
+    pub fn stable_entries(&mut self) -> Vec<Entry> {
+        if self.entries.is_empty() {
+            return Vec::new();
         }
+        self.offset = self.entries.last().unwrap().get_index() + 1;
+        std::mem::take(&mut self.entries)
     }
 
-    /// Removes the snapshot from self if the index of the snapshot matches
-    pub fn stable_snap_to(&mut self, idx: u64) {
-        if self.snapshot.is_none() {
-            return;
-        }
-        if idx == self.snapshot.as_ref().unwrap().get_metadata().index {
-            self.snapshot = None;
-        }
+    /// Returns the snapshot and removes it from self.
+    pub fn stable_snap(&mut self) -> Option<Snapshot> {
```

Comment:
This can cause the `first_index` go backward in the similar case as `stable_entries`.